package zurg;

option cc_generic_services = true;
option java_generic_services = true;
option py_generic_services = true;
option java_package = "com.chenshuo.muduo.zurg.proto";
option java_outer_classname = "SlaveProto";

message GetFileContentRequest {
  required string file_name = 1;
  optional int32 max_size = 2 [default = 1048576];
}

message GetFileContentResponse {
  required int32 error_code = 1;
  optional bytes content = 2;
  optional int64 file_size = 3;
}

message RunCommandRequest {
  required string command = 1;
  optional string cwd = 2 [default = "/tmp"];
  repeated string args = 3;
  repeated string envs = 4;
  optional bool envs_only = 5 [default = false];
  optional int32 max_stdout = 6 [default = 1048576];
  optional int32 max_stderr = 7 [default = 1048576];
  optional int32 timeout = 8 [default = 60];
  optional int32 max_memory_mb = 9 [default = 32768];
}

message RunCommandResponse {
  required int32 error_code = 1;
  optional int32 pid = 2;
  optional int32 status = 3;
  optional bytes std_output = 4;
  optional bytes std_error = 5;

  optional int64 start_time_us  = 16;
  optional int64 finish_time_us = 17;
  optional float user_time      = 18;
  optional float system_time    = 19;

  optional int64 memory_maxrss_kb = 20;
  // optional int64 ctxsw = 11;

  optional int32 exit_status = 30 [default = 0];
  optional int32 signaled = 31 [default = 0];
  optional bool coredump = 32 [default = false];
}

service SlaveService {
  rpc getFileContent (GetFileContentRequest) returns (GetFileContentResponse);
  rpc runCommand (RunCommandRequest) returns (RunCommandResponse);
}

